File: samples/nodejs/metadata_cache.js
//
// metadata_cache.js — companion to ../../metadata-cache.md
//
// The JS twin of ../cpp/metadata_cache.cpp: the metadata cache's visibility
// rule observed through two wire-protocol attachments.  An uncommitted
// ALTER is visible only to its own transaction; a committed one is visible
// immediately even to a statement prepared inside an older, still-open
// SNAPSHOT transaction (metadata is read-committed, records are not); two
// uncommitted DDLs collide with the engine's verbatim newVersion error; and
// RDB$FORMATS keeps one row per shape the table has lived through.
//
// Run: node metadata_cache.js
//
'use strict';

const { attachOrCreate, attach, ISOLATION, s } = require('./common');

const DB = '/tmp/fbhandson/mdc_js.fdb';

async function show(who, q, promise) {
    try {
        const rows = await promise;
        const v = rows && rows[0] ? Object.values(rows[0])[0] : '(ok)';
        console.log(`${who}: ${q} -> ${v === null ? '<null>' : s(v)}`);
    } catch (e) {
        console.log(`${who}: ${q} -> ERROR: ${e.message.replace(/\n/g, ' ')}`);
    }
}

(async () => {
    const A = await attachOrCreate({ database: DB });
    const B = await attach({ database: DB });
    await A.query('recreate table t (a integer)');
    await A.query('insert into t values (1)');

    console.log('== 1. uncommitted ALTER: visible to creator only ==');
    const aDdl = await A.transaction();          // A's DDL stays uncommitted
    await aDdl.query('alter table t add e integer');
    await show('A (same tx)  ', 'select e from t', aDdl.query('select e from t'));
    await show('B            ', 'select e from t', B.query('select e from t'));

    console.log("\n== 2. committed ALTER: seen even inside B's open SNAPSHOT tx ==");
    const bSnap = await B.transaction(ISOLATION.SNAPSHOT);
    await show('B (snapshot) ', 'select count(*) from t', bSnap.query('select count(*) from t'));
    await aDdl.commit();                         // E becomes committed
    await A.query('alter table t add d integer');// D committed after B's snapshot
    await show('B (same  tx) ', 'select d from t', bSnap.query('select d from t'));
    console.log('   (metadata is read-committed; the statement was prepared\n' +
        "    against the version chain's current head)");
    await bSnap.commit();

    console.log('\n== 3. two uncommitted DDLs on one object ==');
    const aDdl2 = await A.transaction();
    await aDdl2.query('alter table t add f integer');
    const bDdl = await B.transaction();
    await show('B            ', 'alter table t add g integer',
        bDdl.query('alter table t add g integer'));
    await bDdl.rollback();
    await aDdl2.rollback();                      // F vanishes with the rollback

    console.log('\n== 4. RDB$FORMATS after the committed DDL ==');
    const [f] = await A.query(
        `select count(*) c from rdb$formats f
           join rdb$relations r on f.rdb$relation_id = r.rdb$relation_id
          where r.rdb$relation_name = 'T'`);
    console.log(`formats stored for T: ${f.C}`);
    const [row] = await A.query('select a, e, d from t');
    console.log(`row written under shape 1 decodes as: a=${row.A}, e=${row.E}, d=${row.D}`);

    await B.detach();
    await A.detach();
    console.log('done.');
})().catch(e => { console.error('ERR:', e.message); process.exit(1); });

File: samples/rust/src/bin/metadata_cache.rs
//! metadata_cache.rs — the metadata-cache scenario in Rust.
//!
//! The rsfbclient twin of ../cpp/metadata_cache.cpp: the cache's
//! visibility rule observed through two attachments.  An uncommitted
//! ALTER is visible only to its own transaction; a committed one is
//! visible immediately even to a statement prepared inside an older,
//! still-open SNAPSHOT transaction (metadata is read-committed, records
//! are not); two uncommitted DDLs collide with the engine's "object in
//! use" newVersion error; and RDB$FORMATS keeps one row per shape the
//! table has lived through.  Explicit SimpleTransaction objects keep each
//! prepare inside the intended transaction (each connection's hidden
//! default transaction would blur exactly the boundaries this sample is
//! about).  See ../../metadata-cache.md.
//!
//! Run (see ../README.md):  cargo run --bin metadata_cache

use fb_handson_rust::connect;
use rsfbclient::{prelude::*, FbError, SimpleTransaction};

fn flat(e: &FbError) -> String {
    e.to_string().replace('\n', " ")
}

// Run a single-value query in the given transaction, printing the value
// or the error (both outcomes are data here).
fn try_query(who: &str, tr: &mut SimpleTransaction, sql: &str) {
    match tr.query_first::<(), (Option<i64>,)>(sql, ()) {
        Ok(Some((Some(v),))) => println!("{}: {} -> {}", who, sql, v),
        Ok(Some((None,))) => println!("{}: {} -> <null>", who, sql),
        Ok(None) => println!("{}: {} -> (no rows)", who, sql),
        Err(e) => println!("{}: {} -> ERROR: {}", who, sql, flat(&e)),
    }
}

fn main() -> Result<(), FbError> {
    let mut a = connect("metadata_cache")?;
    let mut b = connect("metadata_cache")?;

    {
        let mut t = SimpleTransaction::new(&mut a, Default::default())?;
        t.execute("recreate table t (a integer)", ())?;
        t.commit()?;
    }
    {
        let mut t = SimpleTransaction::new(&mut a, Default::default())?;
        t.execute("insert into t values (1)", ())?;
        t.commit()?;
    }

    // -- 1. uncommitted DDL: mine, and mine alone -----------------------
    println!("== 1. uncommitted ALTER: visible to creator only ==");
    let mut a_ddl = SimpleTransaction::new(&mut a, Default::default())?; // stays open
    a_ddl.execute("alter table t add e integer", ())?;
    try_query("A (same tx)  ", &mut a_ddl, "select e from t");
    {
        let mut b_tra = SimpleTransaction::new(&mut b, Default::default())?;
        try_query("B            ", &mut b_tra, "select e from t");
        b_tra.commit()?;
    }

    // -- 2. committed DDL ignores open snapshots ------------------------
    println!("\n== 2. committed ALTER: seen even inside B's open SNAPSHOT tx ==");
    let snapshot = TransactionConfiguration {
        isolation: TrIsolationLevel::Concurrency,
        ..TransactionConfiguration::default()
    };
    let mut b_snap = SimpleTransaction::new(&mut b, snapshot)?;
    try_query("B (snapshot) ", &mut b_snap, "select count(*) from t");
    a_ddl.commit()?; // E becomes committed
    {
        let mut t = SimpleTransaction::new(&mut a, Default::default())?;
        t.execute("alter table t add d integer", ())?;
        t.commit()?; // D committed after B's snapshot started
    }
    try_query("B (same  tx) ", &mut b_snap, "select d from t");
    println!(
        "   (records are snapshot-isolated; metadata is read-committed —\n\
         \x20   the new statement was prepared against the chain's current head)"
    );
    b_snap.commit()?;

    // -- 3. concurrent DDL: the newVersion collision --------------------
    println!("\n== 3. two uncommitted DDLs on one object ==");
    let mut a_ddl = SimpleTransaction::new(&mut a, Default::default())?;
    a_ddl.execute("alter table t add f integer", ())?;
    let mut b_ddl = SimpleTransaction::new(&mut b, Default::default())?;
    match b_ddl.execute("alter table t add g integer", ()) {
        Ok(_) => println!("B: ALTER unexpectedly succeeded"),
        Err(e) => println!("B: ALTER failed:\n{}", e),
    }
    b_ddl.rollback()?;
    a_ddl.rollback()?; // F vanishes with the rollback

    // -- 4. the on-disk half: one format per committed shape ------------
    println!("\n== 4. RDB$FORMATS after the committed DDL ==");
    let mut t = SimpleTransaction::new(&mut a, Default::default())?;
    let (formats,): (i64,) = t
        .query_first(
            "select count(*) from rdb$formats f \
             join rdb$relations r on f.rdb$relation_id = r.rdb$relation_id \
             where r.rdb$relation_name = 'T'",
            (),
        )?
        .unwrap();
    println!("formats stored for T: {} (T has lived through that many shapes)", formats);
    let (ca, ce, cd): (Option<i64>, Option<i64>, Option<i64>) = t
        .query_first("select a, e, d from t", ())?
        .unwrap();
    let show = |v: Option<i64>| v.map_or("<null>".to_string(), |x| x.to_string());
    println!(
        "row written under shape 1 decodes as: a={}, e={}, d={}",
        show(ca), show(ce), show(cd)
    );
    t.commit()?;

    println!("done.");
    Ok(())
}
